File: protocol.py
"""
LangGraph IDE — Frozen Contract · Python mirror
============================================================================
This file mirrors protocol.ts. Fields, discriminators, and enum values must
match verbatim. The contract round-trip test (test-plan §L1) feeds the same
golden JSON to both sides.

PIN notes (same as protocol.ts): anything marked `# PIN` has a shape that must
be calibrated against a real runtime dump (engineering-design §1).
Depends on: pydantic v2.
"""
from __future__ import annotations

from typing import Annotated, Any, Final, Literal, Union
from pydantic import BaseModel, Field, TypeAdapter

PROTOCOL_VERSION: Final = "0.1.0"

CheckpointId = str
ThreadId = str
RunId = str
# PIN: parallel interrupt id collision (langgraph #6626) -> dedup by node+seq.
# PIN-cal DEFER 2026-06-17: parallel-interrupt behavior on langgraph 1.1.9 not yet
# exercised (no vscode.lm offline); dedup layer lands PHASE 2. See pin_dump.golden.txt.
InterruptId = str
NodeName = str

BoundaryWhen = Literal["before", "after"]
ProviderMode = Literal["copilot", "api", "manual"]
InferenceExpectation = Literal["text", "tool_call"]
TokenSource = Literal["vscode_lm_counttokens", "api_usage", "sidecar_estimate"]


class Envelope(BaseModel):
    v: Literal["0.1.0"] = PROTOCOL_VERSION
    corr: str | None = None


# ---- Token measurement (R3) -----------------------------------------------
class TokenCount(BaseModel):
    prompt: int
    completion: int | None = None
    source: TokenSource


# ---- State (R1/R2) ---------------------------------------------------------
class StateDiffEntry(BaseModel):
    # PIN-cal confirmed 2026-06-17: get_state().values is keyed by channel name
    # (golden: 'messages', 'steps').
    channel: str
    before: Any | None = None
    after: Any | None = None
    op: Literal["add", "update", "remove"]


class StateSnapshot(BaseModel):
    # PIN-cal confirmed 2026-06-17: get_state().values is a dict[str, Any] keyed by
    # channel (pin_dump.golden.txt).
    values: dict[str, Any]
    diff: list[StateDiffEntry] | None = None


# ---- Helpers ---------------------------------------------------------------
class ChatMessage(BaseModel):
    role: Literal["system", "human", "ai", "tool"]
    content: str
    name: str | None = None
    # PIN-cal PENDING 2026-06-17: needs a tool-calling model; offline fake model emits no
    # tool calls. Confirm on PHASE 2 (manual tool_call path / vscode.lm).
    toolCallId: str | None = None


class JsonSchema(BaseModel):
    type: Literal["object"]
    properties: dict[str, dict[str, Any]]
    required: list[str] | None = None


ErrorCode = Literal[
    "consent_denied", "quota_exceeded", "model_unavailable",
    "resume_kind_mismatch", "tool_schema_validation", "interrupt_id_conflict",
    "graph_load_failed", "checkpoint_not_found", "internal",
]


# ============================================================================
# ServerEvent — sidecar -> extension
# ============================================================================
class GraphTopology(Envelope):
    # R1 execution view: get_graph() nodes/edges for the canvas. Added PHASE 1.
    type: Literal["graph"] = "graph"
    threadId: ThreadId | None = None
    nodes: list[NodeName]
    edges: list[tuple[NodeName, NodeName]]
    # JSON Schema of the graph's input (get_input_jsonschema) for the run-input form;
    # None if introspection failed. Added run-input-form.
    inputSchema: dict[str, Any] | None = None
    # Absolute project root the graph was loaded from, so the form can pre-fill
    # path-like inputs (repo_path -> root, out_dir -> root/out). Added ui-design-pass.
    projectRoot: str | None = None
    # First docstring line per node (the node's purpose) for the overview table;
    # None per node if it has no docstring. Added graph-overview-lanes.
    nodeDocs: dict[str, str | None] | None = None
    # Static lane classification per node: "llm" (references a model / calls
    # interrupt) or "script". Best-effort; the webview refines it from runtime
    # llm_start / manual_inference_required events. Added graph-overview-lanes.
    nodeKinds: dict[str, str] | None = None
    # Branch condition per conditional edge, keyed "src->tgt" (e.g. "gate->review":
    # "human"). From get_graph().edges[i].data. Added graph-autolayout.
    edgeLabels: dict[str, str] | None = None


class RunStarted(Envelope):
    type: Literal["run_started"] = "run_started"
    threadId: ThreadId
    runId: RunId
    checkpointId: CheckpointId | None = None


class NodeStart(Envelope):
    type: Literal["node_start"] = "node_start"
    threadId: ThreadId
    runId: RunId
    node: NodeName
    checkpointId: CheckpointId
    ts: float


class NodeEnd(Envelope):
    type: Literal["node_end"] = "node_end"
    threadId: ThreadId
    runId: RunId
    node: NodeName
    checkpointId: CheckpointId
    durationMs: float
    diff: list[StateDiffEntry] | None = None


class LlmStart(Envelope):
    type: Literal["llm_start"] = "llm_start"
    threadId: ThreadId
    runId: RunId
    node: NodeName
    llmEventId: str
    # PIN-cal confirmed 2026-06-17: on_chat_model_start.metadata carries
    # ls_model_type / ls_provider (pin_dump.golden.txt).
    model: str | None = None
    promptTokens: TokenCount | None = None
    # The actual prompt text sent to the model (clipped). Added llm-prompt-view.
    promptText: str | None = None


class LlmToken(Envelope):
    type: Literal["llm_token"] = "llm_token"
    llmEventId: str
    delta: str


class LlmEnd(Envelope):
    type: Literal["llm_end"] = "llm_end"
    llmEventId: str
    tokens: TokenCount | None = None
    finishReason: str | None = None
    # The model's response text (clipped). Added llm-prompt-view.
    completionText: str | None = None


class ToolStart(Envelope):
    type: Literal["tool_start"] = "tool_start"
    threadId: ThreadId
    runId: RunId
    node: NodeName
    toolEventId: str
    name: str
    args: dict[str, Any]


class ToolEnd(Envelope):
    type: Literal["tool_end"] = "tool_end"
    toolEventId: str
    ok: bool
    result: Any | None = None
    error: str | None = None


class ManualInferenceRequired(Envelope):
    """R4 differentiator. expects + toolSchema = the contract answer to open question #2."""
    type: Literal["manual_inference_required"] = "manual_inference_required"
    threadId: ThreadId
    runId: RunId
    node: NodeName
    interruptId: InterruptId
    renderedText: str
    messages: list[ChatMessage]
    expects: InferenceExpectation
    toolSchema: JsonSchema | None = None
    promptTokens: TokenCount


class BreakpointHit(Envelope):
    type: Literal["breakpoint_hit"] = "breakpoint_hit"
    threadId: ThreadId
    runId: RunId
    node: NodeName
    when: BoundaryWhen
    checkpointId: CheckpointId


class StateSnapshotEvent(Envelope):
    type: Literal["state_snapshot"] = "state_snapshot"
    threadId: ThreadId
    checkpointId: CheckpointId
    snapshot: StateSnapshot


class RunFinished(Envelope):
    type: Literal["run_finished"] = "run_finished"
    threadId: ThreadId
    runId: RunId
    status: Literal["completed", "interrupted", "error", "aborted"]
    checkpointId: CheckpointId | None = None


class ErrorEvent(Envelope):
    type: Literal["error"] = "error"
    code: ErrorCode
    message: str
    detail: Any | None = None
    node: NodeName | None = None
    runId: RunId | None = None


ServerEvent = Annotated[
    Union[
        GraphTopology, RunStarted, NodeStart, NodeEnd, LlmStart, LlmToken, LlmEnd,
        ToolStart, ToolEnd, ManualInferenceRequired, BreakpointHit,
        StateSnapshotEvent, RunFinished, ErrorEvent,
    ],
    Field(discriminator="type"),
]


# ============================================================================
# ClientCommand — extension -> sidecar
# ============================================================================
class StartRun(Envelope):
    type: Literal["start_run"] = "start_run"
    threadId: ThreadId | None = None
    input: dict[str, Any]
    providerMode: ProviderMode


class ResumeText(BaseModel):
    kind: Literal["text"] = "text"
    text: str


class ResumeToolCall(BaseModel):
    kind: Literal["tool_call"] = "tool_call"
    name: str
    args: dict[str, Any]


ResumePayload = Annotated[Union[ResumeText, ResumeToolCall], Field(discriminator="kind")]


class Resume(Envelope):
    type: Literal["resume"] = "resume"
    threadId: ThreadId
    interruptId: InterruptId
    payload: ResumePayload


class SetBreakpoint(Envelope):
    type: Literal["set_breakpoint"] = "set_breakpoint"
    node: NodeName
    when: BoundaryWhen


class ClearBreakpoint(Envelope):
    type: Literal["clear_breakpoint"] = "clear_breakpoint"
    node: NodeName
    when: BoundaryWhen


class Step(Envelope):
    type: Literal["step"] = "step"
    threadId: ThreadId
    runId: RunId


class Fork(Envelope):
    type: Literal["fork"] = "fork"
    threadId: ThreadId
    checkpointId: CheckpointId
    stateOverride: dict[str, Any] | None = None


class GetState(Envelope):
    type: Literal["get_state"] = "get_state"
    threadId: ThreadId
    checkpointId: CheckpointId | None = None


class Cancel(Envelope):
    type: Literal["cancel"] = "cancel"
    threadId: ThreadId
    runId: RunId


ClientCommand = Annotated[
    Union[
        StartRun, Resume, SetBreakpoint, ClearBreakpoint,
        Step, Fork, GetState, Cancel,
    ],
    Field(discriminator="type"),
]


# Runtime parse helpers (Python-only; the TS mirror uses zod discriminated unions).
ServerEventAdapter: TypeAdapter[ServerEvent] = TypeAdapter(ServerEvent)
ClientCommandAdapter: TypeAdapter[ClientCommand] = TypeAdapter(ClientCommand)

File: protocol.ts
/**
 * LangGraph IDE — Frozen Contract · TypeScript mirror
 * ===========================================================================
 * Mirror of protocol.py. Fields, discriminators and enum values must match
 * verbatim; the L1 round-trip test (test/protocol.l1.test.ts + tests/test_protocol_l1.py)
 * feeds the same golden JSON to both sides. Change one side -> change the other + golden.
 *
 * Runtime validation via zod (the Python side uses pydantic). `# PIN` shapes on the
 * Python side are calibrated against pin_dump.golden.txt.
 */
import { z } from "zod";

export const PROTOCOL_VERSION = "0.1.0" as const;

// Envelope fields shared by every wire message.
const env = { v: z.literal(PROTOCOL_VERSION), corr: z.string().nullable() };

// ---- enums -----------------------------------------------------------------
export const BoundaryWhen = z.enum(["before", "after"]);
export const ProviderMode = z.enum(["copilot", "api", "manual"]);
export const InferenceExpectation = z.enum(["text", "tool_call"]);
export const TokenSource = z.enum(["vscode_lm_counttokens", "api_usage", "sidecar_estimate"]);
export const ErrorCode = z.enum([
  "consent_denied", "quota_exceeded", "model_unavailable",
  "resume_kind_mismatch", "tool_schema_validation", "interrupt_id_conflict",
  "graph_load_failed", "checkpoint_not_found", "internal",
]);

// ---- helpers ---------------------------------------------------------------
export const TokenCount = z.object({
  prompt: z.number().int(),
  completion: z.number().int().nullable(),
  source: TokenSource,
});
export const StateDiffEntry = z.object({
  channel: z.string(),
  before: z.any().nullable(),
  after: z.any().nullable(),
  op: z.enum(["add", "update", "remove"]),
});
export const StateSnapshot = z.object({
  values: z.record(z.any()),
  diff: z.array(StateDiffEntry).nullable(),
});
export const ChatMessage = z.object({
  role: z.enum(["system", "human", "ai", "tool"]),
  content: z.string(),
  name: z.string().nullable(),
  toolCallId: z.string().nullable(),
});
export const JsonSchema = z.object({
  type: z.literal("object"),
  properties: z.record(z.record(z.any())),
  required: z.array(z.string()).nullable(),
});
// Where a node's function is defined, for jump-to-source (P1-1). file = absolute path,
// line = 1-based def line.
export const SourceRef = z.object({ file: z.string(), line: z.number().int() });

// ============================================================================
// ServerEvent — sidecar -> extension
// ============================================================================
export const GraphTopology = z.object({
  ...env, type: z.literal("graph"),
  threadId: z.string().nullable(),
  nodes: z.array(z.string()),
  edges: z.array(z.tuple([z.string(), z.string()])),
  inputSchema: z.record(z.any()).nullable(),
  projectRoot: z.string().nullable().default(null),
  nodeDocs: z.record(z.string().nullable()).nullable().default(null),
  nodeKinds: z.record(z.string()).nullable().default(null),
  edgeLabels: z.record(z.string()).nullable().default(null),
  nodeSources: z.record(SourceRef).nullable().default(null),
  hasCheckpointer: z.boolean().nullable().default(null),
  langgraphVersion: z.string().nullable().default(null),
  workerPython: z.string().nullable().default(null),
});
export const RunStarted = z.object({
  ...env, type: z.literal("run_started"),
  threadId: z.string(), runId: z.string(), checkpointId: z.string().nullable(),
});
export const NodeStart = z.object({
  ...env, type: z.literal("node_start"),
  threadId: z.string(), runId: z.string(), node: z.string(), checkpointId: z.string(), ts: z.number(),
});
export const NodeEnd = z.object({
  ...env, type: z.literal("node_end"),
  threadId: z.string(), runId: z.string(), node: z.string(), checkpointId: z.string(),
  durationMs: z.number(), diff: z.array(StateDiffEntry).nullable(),
});
export const LlmStart = z.object({
  ...env, type: z.literal("llm_start"),
  threadId: z.string(), runId: z.string(), node: z.string(), llmEventId: z.string(),
  model: z.string().nullable(), promptTokens: TokenCount.nullable(),
  promptText: z.string().nullable().default(null),
});
export const LlmToken = z.object({
  ...env, type: z.literal("llm_token"), llmEventId: z.string(), delta: z.string(),
});
export const LlmEnd = z.object({
  ...env, type: z.literal("llm_end"), llmEventId: z.string(),
  tokens: TokenCount.nullable(), finishReason: z.string().nullable(),
  completionText: z.string().nullable().default(null),
});
export const ToolStart = z.object({
  ...env, type: z.literal("tool_start"),
  threadId: z.string(), runId: z.string(), node: z.string(), toolEventId: z.string(),
  name: z.string(), args: z.record(z.any()),
});
export const ToolEnd = z.object({
  ...env, type: z.literal("tool_end"), toolEventId: z.string(),
  ok: z.boolean(), result: z.any().nullable(), error: z.string().nullable(),
});
export const ManualInferenceRequired = z.object({
  ...env, type: z.literal("manual_inference_required"),
  threadId: z.string(), runId: z.string(), node: z.string(), interruptId: z.string(),
  renderedText: z.string(), messages: z.array(ChatMessage),
  expects: InferenceExpectation, toolSchema: JsonSchema.nullable(), promptTokens: TokenCount,
});
export const BreakpointHit = z.object({
  ...env, type: z.literal("breakpoint_hit"),
  threadId: z.string(), runId: z.string(), node: z.string(), when: BoundaryWhen, checkpointId: z.string(),
});
export const CheckpointRef = z.object({
  checkpointId: z.string(),
  node: z.string().nullable(),  // next node to run from here ("rewind to before <node>"); null at the end
});
export const StateSnapshotEvent = z.object({
  ...env, type: z.literal("state_snapshot"),
  threadId: z.string(), checkpointId: z.string(), snapshot: StateSnapshot,
});
export const CheckpointHistory = z.object({
  ...env, type: z.literal("checkpoint_history"),
  threadId: z.string(), checkpoints: z.array(CheckpointRef),
});
// One conditional-edge (router) decision, reconstructed from the checkpoint history (P1-3):
// at `source` the router chose `key` -> `target`; `alternatives` is the full {key: target} map.
export const BranchDecision = z.object({
  source: z.string(),
  key: z.string().nullable(),
  target: z.string(),
  alternatives: z.record(z.string()),
  stateValues: z.record(z.any()),
});
export const BranchDecisions = z.object({
  ...env, type: z.literal("branch_decisions"),
  threadId: z.string(), decisions: z.array(BranchDecision),
});
export const RunFinished = z.object({
  ...env, type: z.literal("run_finished"),
  threadId: z.string(), runId: z.string(),
  status: z.enum(["completed", "interrupted", "error", "aborted"]), checkpointId: z.string().nullable(),
});
export const ErrorEvent = z.object({
  ...env, type: z.literal("error"), code: ErrorCode, message: z.string(),
  detail: z.any().nullable(), node: z.string().nullable(), runId: z.string().nullable(),
});

export const ServerEvent = z.discriminatedUnion("type", [
  GraphTopology, RunStarted, NodeStart, NodeEnd, LlmStart, LlmToken, LlmEnd,
  ToolStart, ToolEnd, ManualInferenceRequired, BreakpointHit,
  StateSnapshotEvent, CheckpointHistory, BranchDecisions, RunFinished, ErrorEvent,
]);
export type ServerEvent = z.infer<typeof ServerEvent>;

// ============================================================================
// ClientCommand — extension -> sidecar
// ============================================================================
export const StartRun = z.object({
  ...env, type: z.literal("start_run"),
  threadId: z.string().nullable(), input: z.record(z.any()), providerMode: ProviderMode,
});
export const ResumeText = z.object({ kind: z.literal("text"), text: z.string() });
export const ResumeToolCall = z.object({
  kind: z.literal("tool_call"), name: z.string(), args: z.record(z.any()),
});
export const ResumePayload = z.discriminatedUnion("kind", [ResumeText, ResumeToolCall]);
export const Resume = z.object({
  ...env, type: z.literal("resume"), threadId: z.string(), interruptId: z.string(), payload: ResumePayload,
});
export const SetBreakpoint = z.object({
  ...env, type: z.literal("set_breakpoint"), node: z.string(), when: BoundaryWhen,
});
export const ClearBreakpoint = z.object({
  ...env, type: z.literal("clear_breakpoint"), node: z.string(), when: BoundaryWhen,
});
export const Step = z.object({ ...env, type: z.literal("step"), threadId: z.string(), runId: z.string() });
export const Fork = z.object({
  ...env, type: z.literal("fork"), threadId: z.string(), checkpointId: z.string(),
  stateOverride: z.record(z.any()).nullable(),
});
export const GetState = z.object({
  ...env, type: z.literal("get_state"), threadId: z.string(), checkpointId: z.string().nullable(),
});
export const Cancel = z.object({ ...env, type: z.literal("cancel"), threadId: z.string(), runId: z.string() });

export const ClientCommand = z.discriminatedUnion("type", [
  StartRun, Resume, SetBreakpoint, ClearBreakpoint, Step, Fork, GetState, Cancel,
]);
export type ClientCommand = z.infer<typeof ClientCommand>;

export const parseServerEvent = (raw: unknown): ServerEvent => ServerEvent.parse(raw);
export const parseClientCommand = (raw: unknown): ClientCommand => ClientCommand.parse(raw);
